lib/ProofSet.js

```diff
@@ -197,38 +199,108 @@ async function _getProofs({document}) {
+  // map each purpose to at least one proof to verify
+  const purposes = Array.isArray(purpose) ? purpose : [purpose];
+  const purposeToProofs = new Map();
+  const proofToSuite = new Map();
+  const suiteMatchQueue = new Map();
+  await Promise.all(purposes.map(purpose => _matchProofSet({
+    purposeToProofs, proofToSuite, purpose, proofSet, suites,
+    suiteMatchQueue, document, documentLoader, expansionMap
+  })));
+
+  // every purpose must have at least one matching proof or verify will fail
+  if(purposeToProofs.length < purposes.length) {
+    // insufficient proofs to verify, so don't bother verifying any
     return [];
   }
 
-  // verify each matching proof
-  return (await Promise.all(matches.map(async proof => {
-    for(const s of suites) {
-      if(await s.matchProof({proof, document, documentLoader, expansionMap})) {
-        return s.verifyProof({
-          proof, document, purpose, documentLoader, expansionMap
-        }).catch(error => ({verified: false, error}));
+  // verify every proof in `proofToSuite`; these proofs matched a purpose
+  const verifyResults = new Map();
+  await Promise.all([...proofToSuite.entries()].map(async ([proof, suite]) => {
+    let result;
+    try {
+      // create backwards-compatible deferred proof purpose to capture
+      // verification method from old-style suites
+      let vm;
+      const purpose = {
+        async validate(proof, {verificationMethod}) {
+          vm = verificationMethod;
+          return {valid: true};
+        }
+      };
+      const {verified, verificationMethod, error} = await suite.verifyProof({
+        proof, document, purpose, documentLoader, expansionMap
+      });
+      if(!vm) {
+        vm = verificationMethod;
       }
+      result = {proof, verified, verificationMethod: vm, error};
+    } catch(error) {
+      result = {proof, verified: false, error};
     }
-  }))).map((r, i) => {
-    if(!r) {
-      return null;
-    }
-    if(r.error) {
-      _addToJSON(r.error);
+
+    if(result.error) {
+      // ensure error is serializable
+      _makeSerializable(result.error);
     }
-    return {proof: matches[i], ...r};
-  }).filter(r => r);
+
+    verifyResults.set(proof, result);
+  }));
+
+  // validate proof against each purpose that matched it
+  await Promise.all([...purposeToProofs.entries()].map(
+    async ([purpose, proofs]) => {
+      for(const proof of proofs) {
+        const result = verifyResults.get(proof);
+        if(!result.verified) {
+          // if proof was not verified, so not bother validating purpose
+          continue;
+        }
+
+        // validate purpose
+        const {verificationMethod} = result;
+        const suite = proofToSuite.get(proof);
+        let purposeResult;
+        try {
+          purposeResult = await purpose.validate(proof, {
+            document, suite, verificationMethod, documentLoader, expansionMap
+          });
+        } catch(error) {
+          purposeResult = {valid: false, error};
```

Comment:
Again, do we need a `_makeSerializable(error)` call here?

Reply:
No, it happens below in one place: https://github.com/digitalbazaar/jsonld-signatures/pull/149/files#diff-01c3bd2f5d6eee0565c498c5535384249141782a41d6b8d5da84a6fc183a96e7R287